vue-cli -> vite

milespetrov opened the pull request:
# webpack/vue-cli -> vitejs

This PR should be locally tested by all team members prior to merging. This PR also blocks all other PR's as rebasing just isn't fun.


[Demo](https://ramp4-app.azureedge.net/demo/users/milespetrov/develop/index.html)

Initial performance (demo vs current master):


|         | Network Requests | Assets size (browser) | Finish load time (browser) | Dev server start time    |
| ------- | ---------------- | --------------------- | -------------------------- | --- |
| Vitejs  | 319              | 9.2 MB                | 5.96s                      |  1s |
| Webpack | 260              | 18.9 MB               | 6.44s                      |  30s|


Important changes:
- Starter scripts are native ESM and contain `type="module"`. 
    - `<script type="module" src="./starter-scripts/main.ts"></script>`
    - [Module Guide](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#applying_the_module_to_your_html)
- Our own starter scripts can be written in TypeScript or JS. External users just adding the above script to their page need to use JS since browsers don't process TS. If they are building a TS project with ramp then their TS compiler should convert any custom TS starter scripts. We should likely provide samples of both JS and TS.
- [Type-Only Imports and Export](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export) are used to avoid errors where the browser tries to import a type (not a JS thing) vs a value.

> import type only imports declarations to be used for type annotations and declarations. It always gets fully erased, so there’s no remnant of it at runtime. Similarly, export type only provides an export that can be used for type contexts, and is also erased from TypeScript’s output.

> It’s important to note that classes have a value at runtime and a type at design-time, and the use is context-sensitive. When using import type to import a class, you can’t do things like extend from it.

Recommended readings if you like build stuff: 
- [Why Vite](https://vitejs.dev/guide/why.html)
- [Vite Features](https://vitejs.dev/guide/features.html)

# How to test

Clone [milespetrov/ramp4-pcar4](https://github.com/milespetrov/ramp4-pcar4.git) into a new folder and checkout the `develop` branch.

Download the latest [Node LTS version](https://nodejs.org/en/download/), currently v16.14.x. Node v17.9.x also works.

If you want to keep v14.15.4 or need to use different node versions between projects checkout [nvm](https://github.com/nvm-sh/nvm) instead.

## Recommended IDE Setup

[VSCode](https://code.visualstudio.com/) with the recommended extensions (VSCode should bug you to install them).

### Important:

1. Install the [Volar](https://marketplace.visualstudio.com/items?itemName=johnsoncodehk.volar) extension.
2. Disable/remove [Vetur](https://marketplace.visualstudio.com/items?itemName=octref.vetur).
3. Type `@builtin typescript` in the search box on the VSCode extensions tab and **disable** "TypeScript and JavaScript Language Features". Volar has its own TS language server so we don't want to run two concurrently.

### Install dependencies
```sh
npm install
```

### Compile and Hot-Reload for Development

```sh
npm run dev
```

Open `http://localhost:3000` in your browser.

### Build for production

```sh
npm run build
```

The production files will be placed in the `dist` folder.

### Preview production build (after running build)

```sh
npm run preview
```

Open `http://localhost:5050` in your browser.

# Todos

1. Only `public/index.html` and `public/starter-scripts/main.ts` are up-to-date. The rest won't be much trouble to update. The only change is to import `RAMP` instead of getting it from the window. 
2. Vue is bundled and `RAMP` is imported from `@/api`. This can be changed to externalize Vue and also to add / change what we want to export to external users.
3. [shpjs](https://www.npmjs.com/package/shpjs) wasn't working well in the browser as ESM. As a temporary workaround I placed a plain script tag to [unpkg.com](https://unpkg.com/shpjs@latest/dist/shp.js) in the index.html file.
4. Verify / fix cypress testing (if we're still doing that). 
5. Verify / fix any unwanted css changes if any. Overall it seems the same as before. 
6. To simplify this build change I didn't add monorepo support, though it would be fairly trivial to add later with the built-in npm workspace support if we want that. The `ramp-locale-loader` has been moved into this repo as a vite plugin.
7. Go though and sanity check everything. Minor things can be addressed in later PR's, major things will be investigated for this PR.

# Some observations for later discussions

- `vuex-pathify` has a `v3.0.0-beta` release supporting Vue 3.
- `pinia` is a great alternative to vuex in regards to TS support and simplicity.
- Some fixtures (legend in particular) have circular dependencies. I fixed any that broke ESM by using asyncComponents but there could be more lurking.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)
<!-- Reviewable:end -->

james-rae commented:
First, :trophy: as this is a big update and is probably the :brain: stack path we want to be pursuing. Also, dang that is fast dev startup. 

> Only public/index.html and public/starter-scripts/main.ts are up-to-date. The rest won't be much trouble to update. The only change is to import RAMP instead of getting it from the window.

I've noticed I no longer can access `RAMP` from my console. The new `main` is now typescript doing an import (i.e. needs build process prior to running). How does a page that has just added the ramp library as a script access it now to spin up new instances or make calls on what used to be `RAMP`? 

> To simplify this build change I didn't add monorepo support, though it would be fairly trivial to add later with the built-in npm workspace support if we want that. The ramp-locale-loader has been moved into this repo as a vite plugin.

Is this a known, or a "I hope so"? I think for "ramp core library" we are ok, but we had some "fixtures as packages" which was looking like a good way to manage fancy, one-off fixtures without them being "core" or "supported".  If this is easy to do as vite plugins then sounds good.

Other than that, I guess we need to spin up a bunch of new issues to make sure the TODOs above don't get forgotten.
- fix tests (we should know our one working test is still able to function)
- fix `shpjs`, maybe fork to our org and update the guts?
- update startup scripts (`new student` tag)

I'll go through the million or so files during the week I guess.

> `http://localhost:3000`

Making me update my bookmarks, how rude.





<!-- Sent from Reviewable.io -->

james-rae commented:
Reviewable disrespecting my edits, added fouth task in that we def want ability to not have Vue pre-bundled or however that works.





<!-- Sent from Reviewable.io -->

spencerwahl commented:
Runs and nothing seems obviously broken that I've clicked on.

I just added your fork as a new remote and switched to your develop in the same workspace so thats a good sign for no headaches after this is pulled.

sharvenp reviewed (commented):
I noticed that for me some of the layers are stuck in a placeholder state in the legend. The layers that are stuck in this state seems somewhat random as I can sometimes get some of these layers to load by refreshing the page (but then the previously loaded layers get stuck). Additionally, any layer added through the wizard also gets stuck, and there are no errors/warnings in the console related to this.

However the layers function properly on the map (identify, reorder, hovertips work) and apart from this the app seems to run okay. 

> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-:-N02Tb5e1YZmbEAT9XTq:bg3upel)* status: 0 of 481 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

RyanCoulsonCA reviewed (commented):
Other than the few issues that have already been brought up, I haven't noticed anything else while playing around with the app. Had no problems with the set up either.

> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-:-N05uBvH8TbUTazqvTP9:b-ihf69k)* status: 0 of 481 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

james-rae reviewed (changes requested):
I see we have `package-lock.json` again. My memories of this file is one of anger lol. Lets come up with some team guidelines for how this file gets treated (similar to our now obsolete rules about when `rush update` is allowed)

As an aside, didn't know about `import` vs `import type` but looks like a good enhancement here.

> Reviewed 481 of 481 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-:-N06xx2FCdYoXq0gdOh-:b-qq3ivf)* status: all files reviewed, 2 unresolved discussions (waiting on @milespetrov)

___
*[.github/workflows/pr-checks.yml, line 5 at r1](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-N06zN-I411_UelfRRPw:-N06zN-I411_UelfRRPx:b-uoddb1) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/aaa25872b96ae608cfc71c2c3e8e20c0fc3aa1c9/.github/workflows/pr-checks.yml#L5)):*
> ```YAML
>         branches: [master]
> env:
>     NODE_VERSION: '14.15.4'
> ```

this number seems out of sync with what got changed in `azure-blob-publish.yml`
___
*[packages/ramp-core/src/fixtures/legend/components/component.vue, line 34 at r1](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-N074QYg8qiLxJKebFYi:-N074QYg8qiLxJKebFYj:b58hmd) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/e79bf4e4790a34d00f3e71645efb6cf00eaa7f27/packages/ramp-core/src/fixtures/legend/components/component.vue#L34)):*
> ```Vue
>     },
>     mounted() {
>         this.legendItem.loadPromise.then(() => {
> ```

this being deleted could be whats causing the legend problems sharven noticed.


<!-- Sent from Reviewable.io -->

james-rae reviewed (changes requested):
> Reviewed 7 of 10 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)* status: 481 of 484 files reviewed, 3 unresolved discussions (waiting on @james-rae and @milespetrov)

___
*[src/fixtures/help/screen.vue, line 7 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-N07kotnEUgZBC5Qjgjs:-N07kotnEUgZBC5Qjgjt:blhptwr) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/41f0fd156e993454caa0fe4c44bf5364c8085f73/src/fixtures/help/screen.vue#L7)):*
> ```Vue
> minor cleanups Latest commit 5612ccd on Feb 10 History 5 contributors
> @yileifeng@RyanCoulsonCA@milespetrov@james-rae@elsa-huang 112 lines (102 sloc)
> 4.15 KB
> ```

lol i don't think you meant for this to be in this file.


<!-- Sent from Reviewable.io -->

james-rae reviewed (commented):
> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)* status: all files reviewed, 2 unresolved discussions (waiting on @milespetrov)



<!-- Sent from Reviewable.io -->

james-rae reviewed (changes requested):
> Reviewed 27 of 27 files at r4.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)* status: 468 of 499 files reviewed, 4 unresolved discussions (waiting on @james-rae and @milespetrov)

___
*[README.md, line 29 at r4](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-N0C6jIQ4Zbz-WvinFQn:-N0C6jIQ4Zbz-WvinFQo:bl7572h) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/0eeed9f1354794c1329a278212dfbaf9a6d6b061/README.md#L29)):*
> ```Markdown
> ```
> 
> Then open `http://localhost:5050/public/external.html` in your brwoser.
> ```

type-o `brwoser`
___
*[public/starter-scripts/main.js, line 56 at r4](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-N0C8PQv8FIZZNzhIbcy:-N0C8PQv8FIZZNzhIbcz:b-4f1xsg) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/0eeed9f1354794c1329a278212dfbaf9a6d6b061/public/starter-scripts/main.js#L56)):*
> ```JavaScript
>                     {
>                         id: 'LOD_NRCAN_Lambert_3978',
>                         lods: RAMP.geoApi.defaultLODs(
> ```

if we're back to using `RAMP.`, is there a reason to change the names? 
I'd be ok if `.GEO.` becomes `.geo.`, we had kept all caps to be same as R2 but not that big of a deal.

`new RAMP.Instance()` seems more descriptive that `RAMP.useRamp()`. If we need functions, call it something like `createX()` or `newX()`, where X could be `Ramp`, `Map`, `App`, `Instance` type of thing.  Open to haggle!


<!-- Sent from Reviewable.io -->

james-rae reviewed (commented):
> Reviewed 19 of 19 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)* status: all files reviewed, 4 unresolved discussions (waiting on @milespetrov)



<!-- Sent from Reviewable.io -->

sharvenp reviewed (commented):
I noticed that opening details for layers with custom templates (WFSLayer and WaterQuantity) throws a Vue warning (can be added as a TODO fix for later):
> The `compilerOptions` config option is only respected when using a build of Vue.js that includes the runtime compiler (aka "full build"). Since you are using the runtime-only build, `compilerOptions` must be passed to `@vue/compiler-dom` in the build setup instead.
> - For vue-loader: pass it via vue-loader's `compilerOptions` loader option.
> - For vue-cli: see https://cli.vuejs.org/guide/webpack.html#modifying-options-of-a-loader
> - For vite: pass it via @vitejs/plugin-vue options. See https://github.com/vitejs/vite/tree/main/packages/plugin-vue#example-for-passing-options-to-vuecompiler-dom

I also tried previewing with build with `npm run build` + `npm run preview` and the page shows a 404 GET request error: `Cannot GET /local/index.html`.

> Reviewed 474 of 481 files at r1, 6 of 10 files at r2, 4 of 6 files at r3, 27 of 27 files at r4, 19 of 19 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-:-N0BiWw08BNrSxsnLwOE:bibeij9)* status: all files reviewed, 4 unresolved discussions (waiting on @milespetrov)



<!-- Sent from Reviewable.io -->

james-rae reviewed (changes requested):
> Reviewed 15 of 37 files at r6, 24 of 24 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)* status: all files reviewed, 2 unresolved discussions (waiting on @milespetrov)

___
*[README.md, line 31 at r7](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962#-N0G4_g22sYCpHUgqrx1:-N0G4_g22sYCpHUgqrx2:b-f07uii) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/72505bf1d0af7f86f52fb7642623900def62f248/README.md#L31)):*
> ```Markdown
> Then open `http://localhost:5050/samples/index.html` in your browser.
> 
> The `demos` folder is _is_ processed by vite, it can reference any file in the repo, and is used during development. This is where our internal development html/js files go. For example, the `demos/starter-scripts/main.js` file imports `{ useRamp, geoApi } from '@/main';` whereas `public/samples/starter-scripts/main.js` doesn't since RAMP is globally defined by the `index.html` file when it loads `<script src="./lib/ramp.iife.js"></script>`.
> ```

` is _is_`, unless you really wanted to emphasize the is-ness


<!-- Sent from Reviewable.io -->

milespetrov commented:
# Final updates

Reviewable is refusing to open for me on this PR, so I'll post updates here.

> I've noticed I no longer can access RAMP from my console. The new main is now typescript doing an import (i.e. needs build process prior to running). How does a page that has just added the ramp library as a script access it now to spin up new instances or make calls on what used to be RAMP? 

Initially I only provided the "production" ESM build which doesn't provide a global variable. 

You can see this type of build [here](https://ramp4-app.azureedge.net/demo/users/milespetrov/develop/demos/index.html).

I've added library builds to the mix (ES, UMD, and IIFE). There's a sample available that uses the `ramp.iife.js` file that has a global `RAMP` and is used in that startup script.

That sample is [here](https://ramp4-app.azureedge.net/demo/users/milespetrov/develop/samples/index.html).

Since we now host both library and production versions I updated the readme, [public/samples vs demo folders](https://github.com/milespetrov/ramp4-pcar4/blob/develop/README.md#local-development).


> Is this a known, or a "I hope so"? I think for "ramp core library" we are ok, but we had some "fixtures as packages" which was looking like a good way to manage fancy, one-off fixtures without them being "core" or "supported". If this is easy to do as vite plugins then sounds good.

It's a fairly confident "known". npm supports workspaces, and vite treats workspace dependencies as source code (rather than ignoring it like node_modules). Unless somethings broken it is well supported. Though it might take time to figure out and learn the little details.


> Runs and nothing seems obviously broken that I've clicked on.
I just added your fork as a new remote and switched to your develop in the same workspace so thats a good sign for no headaches after this is pulled.

> Other than the few issues that have already been brought up, I haven't noticed anything else while playing around with the app. Had no problems with the set up either.

🎉🎉🎉


> I noticed that for me some of the layers are stuck in a placeholder state in the legend. The layers that are stuck in this state seems somewhat random as I can sometimes get some of these layers to load by refreshing the page (but then the previously loaded layers get stuck). Additionally, any layer added through the wizard also gets stuck, and there are no errors/warnings in the console related to this.

> this being deleted could be whats causing the legend problems sharven noticed.

Added that back, just in case.

## Fixed the help section not loading

Prettier was removing very important line breaks...

## Updated the public api

Let me know if everyone is ok with the names:

```
{ geo, version, configUpgrade, createInstance}
```

## Anything else?

I disabled typescript checking for builds, and vite doesn't validate TS during serve. We need to go through all the TS and ESLint issues before 1.0 release and update the build pipeline to run those checks.

Let me know if I missed anything or need more clarification.

james-rae reviewed (approved):
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @milespetrov)



<!-- Sent from Reviewable.io -->

james-rae reviewed (approved):
> Reviewed 1 of 1 files at r9, 2 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @milespetrov)



<!-- Sent from Reviewable.io -->

RyanCoulsonCA reviewed (approved):
> Reviewed 440 of 481 files at r1, 1 of 10 files at r2, 2 of 27 files at r4, 2 of 19 files at r5, 13 of 37 files at r6, 23 of 24 files at r7, 1 of 1 files at r9, 2 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @milespetrov)



<!-- Sent from Reviewable.io -->

sharvenp reviewed (approved):
> Reviewed 13 of 37 files at r6, 23 of 24 files at r7, 1 of 1 files at r9, 2 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/962)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @milespetrov)



<!-- Sent from Reviewable.io -->
